drivers: systick: Assert Cortex-M SysTick calculations are valid

gramsay0 opened the pull request:
If `CONFIG_SYS_CLOCK_TICKS_PER_SEC` is set too low then the calculations may overflow.

For example `mps2_an385` 24bit 25MHz counter wraps every 2^24/25000000 = 0.67 seconds.
Setting `CONFIG_SYS_CLOCK_TICKS_PER_SEC=1` means 1 tick is 1.5 wraps of the counter, `MAX_TICKS` becomes -1.

Assert that `MAX_TICKS` is greater than zero.

Fixes: #66578

gramsay0 commented:
That's weird, the `vmu_rt1170` board failed this assertion...

It seems to have a 1GHz clock with default `CONFIG_SYS_CLOCK_TICKS_PER_SEC=100` which results in `MAX_TICKS` = 0.

I wonder if the systick is working properly on that board?

peter-mitsis reviewed (dismissed):
I think this is a good approach.

It is worth noting that  the benchmark test _tests/benchmark/app_kernel/_ will likely need updating as it sets the system tick rate to 1 and has the mps2_an385 (an affected platform) listed as an _integration_platform_ in its _testcase.yaml_.

andyross commented on drivers/timer/cortex_m_systick.c:
```diff
@@ -43,6 +43,13 @@ static uint32_t last_load;
 #endif
 
+/* Build time or runtime assert depending on
+ * CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME.
```
The only driver that supports that runtime weirdness (which is a historical wart we should really remove) is HPET, this seems like a noop here?

gramsay0 replied:
Currently that config is also used for SiLabs Gecko BURTC timer, Renesas RA4M1 SoC and a few other places I haven't looked into.

Renesas RA4M1 is a Cortex-M SoC that uses this systick driver. It would need to be modified to not use `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` before this could be a build time only assert.

I **_think_** the reason those drivers use `CONFIG_TIMER_READS_ITS_FREQUENCY_AT_RUNTIME` is because they specify the system clock prescaler in the device tree (e.g. commit where Renesas RA4M1 added clock control https://github.com/zephyrproject-rtos/zephyr/commit/1741b3a356fadfb01f1375b08cadb7b8e395bfb4).

I noticed stm32 had a reasonable workaround for that:

```
config SYS_CLOCK_TICKS_PER_SEC
	default 4096 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" < 16
	default 2048 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 16
	default 1024 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 32
	default  512 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 64
	default  256 if "$(dt_node_int_prop_int,$(DT_STM32_LPTIM_PATH),$(DT_ST_PRESCALER))" = 128
	depends on STM32_LPTIM_TIMER && STM32_LPTIM_CLOCK_LSE
```

Which seems nicer than having to calculate all timeouts at runtime.

What are your thoughts on this?

gramsay0 commented:
@peter-mitsis @andyross thanks for the feedback. I'll be back from holiday in a few days and take a proper look

This should not be merged as-is, as @peter-mitsis pointed out likely CI failures I missed

gramsay0 committed:
tests/benchmarks: Prevent Cortex-M systick overflowing in tests

Previously there had been a lot of special casing for each Cortex-M board.
mps2_an385 (an integration platform board) was missed.

Fix with a more generic Kconfig solution and remove special cases

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 committed:
drivers: systick: Assert Cortex-M SysTick calculations are valid

If CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC is too high or
CONFIG_SYS_CLOCK_TICKS_PER_SEC is too low then the
calculations may overflow.

For example mps2_an385 24bit 25MHz counter wraps every
2^24/25000000 = 0.67 seconds.
Setting CONFIG_SYS_CLOCK_TICKS_PER_SEC=1 means 1 tick is 1.5
wraps of the counter, MAX_TICKS becomes -1.

Assert that MAX_TICKS is greater than zero.

For a ticked kernel, MAX_TICKS is not used,
ensure that CYC_PER_TICK < COUNTER_MAX.

Signed-off-by: Grant Ramsay <gramsay@enphaseenergy.com>

gramsay0 commented:
@peter-mitsis 

> It is worth noting that the benchmark test _tests/benchmark/app_kernel/_ will likely need updating as it sets the system tick rate to 1 and has the mps2_an385 (an affected platform) listed as an _integration_platform_ in its _testcase.yaml_.

When fixing the tests I noticed a lot of existing special cases for Cortex-M systick.
I had a go at consolidating into a more generic config

I'm happy to modify/undo those changes if anyone disagrees with that approach

gramsay0 commented:
@peter-mitsis @andyross any further comments/suggestions on this?

henrikbrixandersen commented:
@gramsay0 Is the DNM label still supposed to be set?

gramsay0 commented:
> @gramsay0 Is the DNM label still supposed to be set?

@henrikbrixandersen no, that was set because of CI failures that have since been fixed

aescolar commented:
@gramsay0 Please note this PR needs rebasing

gramsay0 commented:
The CI failure seems unrelated, and builds ok on my machine:
```
west build -p -b lpcxpresso55s69/lpc55s69/cpu0/ns tests/benchmarks/latency_measure -T benchmark.kernel.latency.timeslicing
```

MaureenHelm commented:
> The CI failure seems unrelated, and builds ok on my machine:
> 
> ```
> west build -p -b lpcxpresso55s69/lpc55s69/cpu0/ns tests/benchmarks/latency_measure -T benchmark.kernel.latency.timeslicing
> ```

I think this was resolved in main. Can you rebase please?
